internal/xds/translator/local_ratelimit.go

```diff
@@ -0,0 +1,299 @@
+// if applicable, and it does not already exist.
+func (*localRateLimit) patchHCM(mgr *hcmv3.HttpConnectionManager, irListener *ir.HTTPListener) error {
+	if mgr == nil {
+		return errors.New("hcm is nil")
+	}
+	if irListener == nil {
+		return errors.New("ir listener is nil")
+	}
+	if !listenerContainsLocalRateLimit(irListener) {
+		return nil
+	}
+
+	// Return early if filter already exists.
+	for _, httpFilter := range mgr.HttpFilters {
+		if httpFilter.Name == localRateLimitFilter {
+			return nil
+		}
+	}
+
+	localRl := &localrlv3.LocalRateLimit{
+		StatPrefix: localRateLimitFilterStatPrefix,
+	}
+
+	localRlAny, err := anypb.New(localRl)
+	if err != nil {
+		return err
+	}
+
+	// The local rate limit filter at the HTTP connection manager level is an
+	// empty filter. The real configuration is done at the route level.
+	// See https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/local_rate_limit_filter
+	filter := &hcmv3.HttpFilter{
+		Name: localRateLimitFilter,
+		ConfigType: &hcmv3.HttpFilter_TypedConfig{
+			TypedConfig: localRlAny,
+		},
+	}
+
+	mgr.HttpFilters = append(mgr.HttpFilters, filter)
+	return nil
+}
+
+func listenerContainsLocalRateLimit(irListener *ir.HTTPListener) bool {
+	if irListener == nil {
+		return false
+	}
+
+	for _, route := range irListener.Routes {
+		if routeContainsLocalRateLimit(route) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func routeContainsLocalRateLimit(irRoute *ir.HTTPRoute) bool {
+	if irRoute == nil || irRoute.RateLimit == nil || irRoute.RateLimit.Local == nil {
+		return false
+	}
+
+	return true
+}
+
+func (*localRateLimit) patchResources(*types.ResourceVersionTable,
+	[]*ir.HTTPRoute) error {
+	return nil
+}
+
+func (*localRateLimit) patchRouteConfig(*routev3.RouteConfiguration, *ir.HTTPListener) error {
+	return nil
+}
+
+func (*localRateLimit) patchRoute(route *routev3.Route, irRoute *ir.HTTPRoute) error {
+	routeAction := route.GetRoute()
+
+	// Return early if no rate limit config exists.
+	if irRoute.RateLimit == nil || irRoute.RateLimit.Local == nil || routeAction == nil {
+		return nil
+	}
+
+	if routeAction.RateLimits != nil {
+		// This should not happen since this is the only place where the rate limit
+		// config is added in a route.
+		return fmt.Errorf(
+			"route already contains rate limit config:  %s",
+			route.Name)
+	}
+
+	local := irRoute.RateLimit.Local
+
+	rateLimits, descriptors, err := buildRouteLocalRateLimits(irRoute.Name, local)
+	if err != nil {
+		return err
+	}
+	routeAction.RateLimits = rateLimits
+
+	filterCfg := route.GetTypedPerFilterConfig()
+	if _, ok := filterCfg[localRateLimitFilter]; ok {
+		// This should not happen since this is the only place where the filter
+		// config is added in a route.
+		return fmt.Errorf(
+			"route already contains local rate limit filter config:  %s",
+			route.Name)
+	}
+
+	localRl := &localrlv3.LocalRateLimit{
```

Comment:
`always_consume_default_token_bucket` is enable?

Reply:
Yes, the default value is true.  
I believe it's reasonable because the route-level bucket is for all traffic going through that route. But we may need to add a line of comment in the API to explain this. WDYT @zirain @arkodg

Reply:
The default value of Envoy is set to true.  I believe we may need to align the behaviors of both the local and global rate limit and set it to false. WDYT @zirain @arkodg

Reply:
I don't know why envoy set default to ture, but I'm +1 on setting it to `false`